src/vmm/src/controller.rs

```diff
@@ -225,4 +227,243 @@ impl VmmController {
+            vcpu_regs.rax = 2;
+            vcpu_regs.rbx = 3;
+            vcpu_regs.rflags = 2;
+            vcpu_fd.set_regs(&vcpu_regs).unwrap();
+        }
+
+        #[cfg(target_arch = "aarch64")]
+        fn config_vcpu(&self, vcpu_fd: &mut VcpuFd, guest_addr: GuestAddress) {
+            // aarch64 specific registry setup.
+            let mut kvi = kvm_bindings::kvm_vcpu_init::default();
+            self.vm.fd().get_preferred_target(&mut kvi).unwrap();
+            vcpu_fd.vcpu_init(&kvi).unwrap();
+
+            let core_reg_base: u64 = 0x6030_0000_0010_0000;
+            let mut mem_size = 0;
+            self.guest_memory
+                .with_regions_mut(|_, region| -> std::result::Result<(), std::io::Error> {
+                    mem_size += region.len();
+                    Ok(())
+                })
+                .unwrap();
+            let mmio_addr: u64 = (guest_addr.0 + mem_size) as u64;
+            vcpu_fd
+                .set_one_reg(core_reg_base + 2 * 32, guest_addr.0 as u64)
+                .unwrap(); // set PC
+            vcpu_fd
+                .set_one_reg(core_reg_base + 2 * 0, mmio_addr)
+                .unwrap(); // set X0
+        }
+
+        fn write_code_to_guest_memory(&self) -> GuestAddress {
+            let (asm_code, guest_addr) = create_guest_workload();
+            let host_addr = self.guest_memory.get_host_address(guest_addr).unwrap();
+            let code_slice: &mut [u8] =
+                unsafe { std::slice::from_raw_parts_mut(host_addr as *mut u8, asm_code.len()) };
+            code_slice.copy_from_slice(&asm_code);
+            guest_addr
+        }
+
+        // Prepare a VM with 1 MB memory and 1 vCPU.
+        // Dirty a page and run the VM.
+        // Return after the first VM exit.
+        fn setup_with_code_and_run(&mut self) {
+            // Fill guest memory with an asm code snippet.
+            let guest_addr = self.write_code_to_guest_memory();
+
+            // Create a vCPU.
+            let mut vcpu_fd = self.vm.fd().create_vcpu(0).unwrap();
+            self.config_vcpu(&mut vcpu_fd, guest_addr);
+
+            // Run the VM. It will trigger a vCPU exit.
+            match vcpu_fd.run().unwrap() {
+                #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+                VcpuExit::Hlt => (),
+                #[cfg(target_arch = "aarch64")]
+                VcpuExit::MmioWrite(_, _) => (),
+                exit_reason => panic!("unexpected exit reason: {:?}", exit_reason),
+            }
+        }
+    }
+
+    impl VmmController {
+        fn setup_vmm_with_code_and_run(&mut self) {
+            self.vmm.lock().unwrap().setup_with_code_and_run()
+        }
+
+        fn set_dirty_page_tracking(&self, track_dirty_pages: bool) {
+            let vmm = self.vmm.lock().unwrap();
+            config_guest_memory(vmm.guest_memory(), track_dirty_pages, vmm.vm.fd());
+        }
+    }
+
+    #[test]
+    fn test_dirty_bitmap() {
```

Comment:
Maybe we should also test the case where the VM ran and dirtied 1 page, but KVM_DIRTY_BITMAP wasn't enabled. So the result should be an empty bitmap.

Reply:
Added. The result is an error (`ENOENT`) because KVM can't find any regions with dirty tracking enabled.